pkg/prom/cleaner.go

```diff
@@ -0,0 +1,258 @@
+package prom
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/grafana/agent/pkg/prom/instance"
+	"github.com/grafana/agent/pkg/prom/wal"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	promwal "github.com/prometheus/prometheus/tsdb/wal"
+)
+
+const (
+	DefaultCleanupAge    = 12 * time.Hour
+	DefaultCleanupPeriod = 30 * time.Minute
+)
+
+var (
+	discoveryError = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "agent_prometheus_cleaner_storage_discovery_error",
+			Help: "Errors encountered discovering local storage paths",
+		},
+		[]string{"storage"},
+	)
+
+	segmentError = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "agent_prometheus_cleaner_segment_error",
+			Help: "Errors encountered finding most recent WAL segments",
+		},
+		[]string{"storage"},
+	)
+
+	managedStorage = promauto.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "agent_prometheus_cleaner_managed_storage",
+			Help: "Number of storage directories associated with managed instances",
+		},
+	)
+
+	abandonedStorage = promauto.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "agent_prometheus_cleaner_abandoned_storage",
+			Help: "Number of storage directories not associated with any managed instance",
+		},
+	)
+)
+
+// Operations that require interacting with the Prometheus Write Ahead Log (WAL)
+// file format. This interface exists to allow easier testing of the cleaner.
+type walOperations interface {
+	// Get the last modified time of the last segment in a WAL
+	lastModified(path string) (time.Time, error)
+}
+
+type walOperationsDefault struct{}
+
+func (w *walOperationsDefault) lastModified(path string) (time.Time, error) {
+	empty := time.Time{}
+
+	existing, err := promwal.Open(nil, path)
+	if err != nil {
+		return empty, err
```

Comment:
Can we wrap these errors to make it more discoverable about which line is failing?

```go
return empty, fmt.Errorf("could not open WAL: %w", err)
```

(ditto for the rest of this function)

Reply:
Sure